### PR TITLE
[DEVOPS-5720] fix: resolving bugs with java and general improvements/stability/error handling fixes

### PR DESCRIPTION
**Pull request by shakefu**
- refactors ci into common action composites for reuse
- tests both inputs and file based versions
- adds full action test for file based versions
- runs full action tests before releasing
- refactors tools for common logging when checking whether a setup should run
- refactors tools to use `core.exportVariable` instead of modifying process.env directly
- adds env var to control synchronous (easier to debug) or parallel (faster) tool setups
- really really strong error handling in parallel tool setups (probably not doing anything)
- refactor index to allow calling directly from tests without subprocesses
- adds a suite of direct/inline tests for index
- fixes unmarked async tests
- switch naive string parsing of java versions to use regexes instead
- implement check for sdkman having noninteractive settings so ci won't hang
- add generic parser to validateVersion calls
- adds tests for sdkman rc parsing
- adds tests for sdkman config check
- adds tests for validateVersion
- adds `npm run test:debug` for extra output during test runs
- adds tests for individual tools invoked directly with the class to improve coverage
- refactor version() to fix bug and improve handling of all edges
- fix edge case with auto-detecting the tool root directories with command -v
- custom handling of java 1.8.0 version strings to translate to 8.0.x to match sdkman
- adds tests for tool registration

### CODE REVIEW

File: .github/actions/action-test/action.yaml [hunk 1]
```diff
@@ -9,6 +14,23 @@ runs:
...
+    - shell: bash
+      run: |
+        echo "java=17.0.2-tem" > .sdkmanrc
+        echo "16.13.2" > .node-version
+        echo "3.10.2" > .python-version
+        echo "1.17.6" > .go-version
+        echo "1.1.5" > .terraform-version
+      if: inputs.use-versions != 'true'
+    - uses: ./
+      if: inputs.use-versions != 'true'
```
Comment (greenkiwi) [inline .github/actions/action-test/action.yaml:33]: @shakefu I decided it was good to add this test so we tested both cases.  

Of course it works here.  :(

**Review by greenkiwi — APPROVED** [review]

**Review by dgschwindturo — APPROVED** [review]

**Review by greenkiwi — APPROVED** [review]

File: java.js [hunk 2]
```diff
@@ -53,27 +54,29 @@ export default class Java extends Tool {
...
-            : versionIdentifier
+    // parseJavaVersionString specially handles version string extraction
+    // because we have to map strings like 1.8.0_282 to 8.0.282 for the actual
+    // SemVer comparison
+    parseJavaVersionString(expected, version) {
+        // Default case for 11.x or 17.x it should match and we're ok
+        let versions = findVersions(version, { loose: true })
+        if (versions && versions[0] == expected) return versions
+
+        // Handle bullshit case
```
Comment (dgschwindturo) [inline java.js:69]: Would be good to tell the reader why these four lines of code are necessary.

**Review by dgschwindturo — CHANGES_REQUESTED** [review]

File: java.js (hunk 2 above)
Comment (shakefu) [reply java.js:69]: Ugh, fine, I _guess_. :joy:

Comment (shakefu) [reply java.js:69]: Ugh fine, I _guess_. 😂

**Review by dgschwindturo — APPROVED** [review]
